src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
+
+                //copy with offset
+                trit_state[HASH_LENGTH - 1] = 0;
+                System.arraycopy(trit_state, 0, trits, pos, HASH_LENGTH);
+
+                //calculate hash again
+                for (int i = byte_state.length; i-- > 0; ) {
+                    byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
+                }
+                keccak.update(byte_state);
+            }
         } catch (DigestException e) {
-            e.printStackTrace();
+            e.printStackTrace(System.err);
             throw new RuntimeException(e);
         }
     }
 
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
-
+        for (int i = offset; i < offset + size; i++) {
+            if (trits[i] < -1 || trits[i] > 1) {
+                throw new IllegalArgumentException("not a trit: " + trits[i]);
+            }
+        }
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        long num;
+        int power = 0;
+        int n = offset + size - 1;
+        for (; n > offset + 1; n--) {
+            switch (trits[n]) {
+                case 0: // a * b
+                    power++;
+                    break;
+                case 1: // a * b + 1
+                    num = 9 + (trits[--n] * 3) + trits[--n];
+                    for (int count = 0; n > offset && count <= 36; count++, power++) {
+                        num = 3 * num + trits[--n];
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 3]).add(BigInteger.valueOf(num));
+                    power = 0;
+                    break;
+                case -1: // a * b - 1
+                    num = 9 - (trits[--n] * 3) - trits[--n];
+                    for (int count = 0; n > offset && count <= 36; count++, power++) {
+                        num = 3 * num - trits[--n];
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 3]).subtract(BigInteger.valueOf(num));
+                    power = 0;
+                    break;
+
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
+        for (; n >= offset; n--) {
+            switch (trits[n]) {
+                case 0:
+                    power++;
+                    break;
+                case 1:
+                    value = value.multiply(RADIX_POWERS[power + 1]).add(BigInteger.ONE);
+                    power = 0;
+                    break;
+                case -1:
+                    value = value.multiply(RADIX_POWERS[power + 1]).subtract(BigInteger.ONE);
+                    power = 0;
+                    break;
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
```

Comment:
There is a leftover 1 or 2 trits that need  processing .. .and also any un multiplies powers that accumulated.